File: concepts/04_collections_and_iteration/07_sorting_stability_and_custom_order/python/test_02_key_failures_and_comparator_contracts.py
"""排序回调、失败与次序契约。

共同问题：排序回调可能调用多少次；回调失败时输入是否改变；
不一致的次序关系是否由运行时修复。
"""

# polyglot-family: collections_and_iteration
# polyglot-concept: sorting_stability_and_custom_order
# polyglot-related: languages/python/builtins/test_023_general_sequence_types.py

from functools import cmp_to_key

import pytest


def test_sorted_key_failure_leaves_the_input_iterable_unchanged():
    values = ["bbb", "stop", "a"]
    calls = []

    def key(value):
        calls.append(value)
        if value == "stop":
            raise ValueError("cannot rank")
        return len(value)

    with pytest.raises(ValueError, match="rank"):
        sorted(values, key=key)

    assert values == ["bbb", "stop", "a"]
    assert calls == ["bbb", "stop"]


def test_cmp_to_key_adapts_a_repeated_comparator_protocol():
    calls = []

    def compare(left, right):
        calls.append((left, right))
        return (left > right) - (left < right)

    assert sorted([3, 1, 2], key=cmp_to_key(compare)) == [1, 2, 3]
    assert calls

    # 普通 key 每个元素只求值一次；cmp_to_key 包装的 comparator 可被重复调用。比较器
    # 必须给出一致次序，排序实现不会验证传递性或替调用方修复矛盾结果。

File: concepts/04_collections_and_iteration/07_sorting_stability_and_custom_order/nodejs/test_02_key_failures_and_comparator_contracts.mjs
// 排序回调、失败与次序契约。
// 共同问题：排序回调可能调用多少次；回调失败时输入是否改变；
// 不一致的次序关系是否由运行时修复。
//
// polyglot-family: collections_and_iteration
// polyglot-concept: sorting_stability_and_custom_order
// polyglot-related: languages/nodejs/language/test_011_arrays_holes_mutation_copying_sorting_and_species.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

test('toSorted 比较器失败时不修改原数组', () => {
  const failure = new Error('cannot rank');
  const values = [3, 1, 2];

  assert.throws(
    () => values.toSorted(() => {
      throw failure;
    }),
    (error) => error === failure,
  );

  assert.deepEqual(values, [3, 1, 2]);
});

test('NaN 比较结果按相等处理且稳定保留输入顺序', () => {
  const values = [
    { id: 'a', group: 1 },
    { id: 'b', group: 1 },
  ];
  let calls = 0;

  const sorted = values.toSorted(() => {
    calls += 1;
    return Number.NaN;
  });

  assert.deepEqual(sorted.map(({ id }) => id), ['a', 'b']);
  assert.ok(calls > 0);

  // comparator 可重复调用；调用次数不属于接口。非传递或非反对称比较器不满足约定，
  // 结果可能因引擎与输入改变，不能通过单次输出把偶然行为写成规则。
});
